sdk/rust/oak_utils/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use std::collections::HashMap;
+use std::fs;
+use std::io;
+use std::path::Path;
+
+#[cfg(test)]
+mod tests;
+
+// Generates Rust `proto` files in a temporary directory using `protoc_rust`,
+// checks previously generated files and updates them if their contents have changed.
+// This is a workaround for `protoc_rust` issue that always updates files, thus provoking
+// recompilation of all dependent targets.
+// https://github.com/rust-lang/cargo/issues/6529
+// Function doesn't support nested directories since `protoc_rust` doesn't generate them.
+pub fn run_protoc_rust(args: protoc_rust::Args) -> io::Result<()> {
+    let out_path = Path::new(args.out_dir);
+
+    // Create a temporary directory.
+    let temp_dir = tempfile::tempdir()?;
+    let temp_path = temp_dir.path();
+
+    // Generate Rust `proto` files in the temporary directory.
+    let mut temp_args = args;
+    temp_args.out_dir = temp_path.to_str().expect("Temporary path error");
+    protoc_rust::run(temp_args)?;
+
+    // Copy changed Rust `proto` files to the `out_path`.
+    copy_changed_files(temp_path, out_path)?;
+
+    Ok(())
+}
+
+// Generates Rust `proto` files in a temporary directory using `run_protoc_rust`,
+// checks previously generated files and updates them if their contents have changed.
+pub fn run_protoc_rust_grpc(args: protoc_rust_grpc::Args) -> io::Result<()> {
+    let out_path = Path::new(args.out_dir);
+
+    // Create a temporary directory.
+    let temp_dir = tempfile::tempdir()?;
+    let temp_path = temp_dir.path();
+
+    // Generate Rust `grpc` files in the temporary directory.
+    let mut temp_args = args;
+    temp_args.out_dir = temp_path.to_str().expect("Temporary path error");
+    protoc_rust_grpc::run(temp_args)?;
+
+    // Copy changed Rust `grpc` files to the `out_path`.
+    copy_changed_files(temp_path, out_path)?;
+
+    Ok(())
+}
+
+// Traverses `dir` and produces a `HashMap` of filenames and their contents.
+fn get_files(dir: &Path) -> HashMap<String, String> {
+    walkdir::WalkDir::new(dir)
+        .into_iter()
+        .filter_map(|entry| entry.ok())
+        .filter(|entry| entry.path().is_file())
+        .map(|entry| {
+            let path = entry.into_path();
+            let content = fs::read_to_string(&path).expect("Read error");
+            let filename = path
+                .file_name()
+                .expect("Filename error")
+                .to_os_string()
+                .into_string()
+                .expect("OsString error");
+            (filename, content)
+        })
+        .collect()
+}
+
+// This function returns a list of files in the `new_dir` that are different from files with
+// same names in the `old_dir`.
+fn get_changed_and_removed_files(old_dir: &Path, new_dir: &Path) -> (Vec<String>, Vec<String>) {
```

Comment:
I think the return type and the logic are getting a bit too complicated.

Perhaps another solution could be:

- collect filenames from both maps in a single set
- iterate over the set, and get the value from both the old and new map (as options)
- collect the old and new option types in a struct like:

  ```rust
  struct FileDiff {
    old_content: Option<String>,
    new_content: Option<String>,
  }
  ```

- add a method on `FileDiff::action()` to figure out what to do about the file, perhaps it should return one value of some enum:

  ```rust
  enum FileAction {
    Ignore,
    Update{ new_content: String },
    Remove`,
  }
  ```

- apply the necessary change based on this enum value in a final step.

WDYT? I think it should be more readable, and kind of fun to implement in a "functional" style, but if you don't think that's worth it feel free to submit as it is (or do it in a future PR).

Reply:
Will be fixed in https://github.com/project-oak/oak/issues/548